docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md

```diff
@@ -0,0 +1,433 @@
+
+The importance of defining a consistent format and style for the code we write arises from two unfortunate realities: code is easier to write than it is to read, but code is read more often than it is written (or rewritten). 
+
+Code is easier to write than read because, when we are writing it, we usually know exactly what we are trying to do and we have the context needed to understand why and how we're doing it. By contrast, neither of these two things are typically true when we are reading code: we often either don't remember or never knew exactly what the author was trying to do when they wrote the code we're reading, and the context necessary to understand the code is invariably also unknown to us or at least not fresh in our minds. 
+
+This means that, when writing code, we need to have a system of rules and guidelines in place to ensure that the code we write is as easy as possible for others (either our future selves or other developers) to read and understand. This is the purpose of code formatting and style conventions.*
+
+*at a higher level, this is also, in part, the purpose of design patterns and application architectures, but that is a topic for a future lesson.
+
+These conventions are particularly important for us at Arcadia, as most of our code is developed collaboratively, which necessarily means that multiple people will read it. In addition, we often intend for the code we produce to be useful and understandable to others outside of Arcadia, and part of that is ensuring our code meets high standards for readability and comprehensibility. 
+
+
+## An introductory example
+As an example to help illustrate why code formatting and style are so important, consider the following code:
+```python
+def qr(a,b,c):
+    return -(b-np.sqrt(b**2-4*a*c))/2/a,(-b-np.sqrt(b**2-4*a*c))/2/a
+```
+It is very hard to read this code, let alone understand what it does.
+
+Now consider the same code written in a more readable style:
+```python
+def calculate_quadratic_roots(quadratic_coeff, linear_coeff, constant_coeff):
+
+    discriminant_square_root = np.sqrt(
+        linear_coeff ** 2 - 4 * quadratic_coeff * constant_coeff
+    )
+
+    root_1 = (-linear_coeff + discriminant_square_root) / (2 * quadratic_coeff)
+    root_2 = (-linear_coeff - discriminant_square_root) / (2 * quadratic_coeff)
+
+    return root_1, root_2
+```
+Notice how this code is easier to read and easier to understand (if still somewhat opaque). Notice that there are several kinds of changes in this version. For one, spaces are used around the operators, which generally makes code easier to read; this is a kind of *formatting convention*. In addition, the variable names are much longer and are more descriptive, which make it easier for reader to infer the meaning of the code; this is a kind of *style convention*. 
```

Comment:
Sometimes people optimize for line counts ("oh I did this thing in a one-liner" type of attitude). And maybe calling that out specifically and saying in cases like this being more understandable is worth the cost of being less concise etc. 

Some languages are more-prone to this type of stuff. Like Ruby has a lot of syntactic sugar that makes it easy to write cute one-liners that are hard to parse

Reply:
Added!